Added profile preview card on hover feature to ActivityPub

minimaluminium opened the pull request:
ref https://linear.app/ghost/issue/PROD-736/show-profile-preview-on-hover

-  Implemented hover-based profile preview when hovering avatar or name
-  Used shadcn HoverCard component for the interaction pattern
-  Fetched account data on hover to show up-to-date profile info

minimaluminium committed:
Installed hover card component in Shade

This component will be used for the profile preview card when hovering

minimaluminium committed:
Removed title attirbute from avatars

As we now display profile preview when avatars are hovered, we do not need to display the user info in generic title attribute

minimaluminium committed:
Updated account query and follow related caches

- added "enabled" option to useAccountForUser() hook, so that it supports on demand requests which is used for profile preview card
- updated follow/unfollow caches, so that they are updated correctly when the follow button is clicked from the profile preview card

minimaluminium committed:
Added a global component for the profile preview card

- this component uses HoverCard component from shadcn under the hood
- the children of the component will be used as a trigger
- when the trigger is hovered, it first makes a request for fetching account data, and then displays the data in the preview
- if the data is not ready yet, the preview card will display skeleton loader

minimaluminium committed:
Used the profile preview card component in the feed as a starting point

minimaluminium commented on apps/admin-x-activitypub/src/components/global/ProfilePreviewHoverCard.tsx:
```diff
@@ -0,0 +1,156 @@
+    const followsYou = Boolean(displayData?.followsMe);
+    const followingCount = typeof displayData?.followingCount === 'number' ? displayData.followingCount : (Number(displayData?.followingCount) || 0);
+    const followerCount = typeof displayData?.followerCount === 'number' ? displayData.followerCount : (Number(displayData?.followerCount) || 0);
+    const bio = displayData?.bio ? stripHtml(displayData.bio, ['a']) : undefined;
```
`bio` is stripped with our custom html strip function that's used throughout the app.

minimaluminium replied:
I think we should address this in a separate issue as it's now starting to get out of context. No need to do anything for now.

minimaluminium commented on apps/shade/src/components/ui/hover-card.tsx:
```diff
@@ -0,0 +1,27 @@
+  React.ElementRef<typeof HoverCardPrimitive.Content>,
+  React.ComponentPropsWithoutRef<typeof HoverCardPrimitive.Content>
+>(({className, align = 'center', sideOffset = 4, ...props}, ref) => (
+    <HoverCardPrimitive.Content
+        ref={ref}
+        align={align}
+        className={cn(
+            'z-50 w-64 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left- ...
+            className
+        )}
+        sideOffset={sideOffset}
+        {...props}
+    />
```
This is the default implementation from shadcn component—I think we should go with this first and see if it causes any issue. We're using this inside a scrollable suggested profiles in the feed, and there's no issue with clipping etc.

minimaluminium committed:
Used a helper to open links in a new tab for the bio

minimaluminium committed:
Added href validation to the new tab helper

This was recommended by coderabbit. As the amount of change was small, including the fix right in the PR instead of creating a new issue

minimaluminium committed:
Put it behind a feature flag "preview"

minimaluminium committed:
Reverted unnecessary cache control update

minimaluminium committed:
Removed version range from react-hover-card

minimaluminium commented on apps/admin-x-activitypub/src/hooks/use-activity-pub-queries.ts:
```diff
@@ -1114,7 +1114,7 @@ export function useUnfollowMutationForUser(handle: string, onSuccess: () => void
 
             // Update suggested profiles cache (for all limit values)
-            queryClient.setQueriesData({queryKey: ['suggested_profiles_json'], exact: false}, (current: Account[] | undefined) => {
+            queryClient.setQueriesData({queryKey: ['suggested_profiles'], exact: false}, (current: Account[] | undefined) => {
```
This bug was found when I added profile preview card to profile suggestion in the feed. When follow button is clicked in the preview card, the follow button in the suggestions wasn't updating, so I had to include this fix in the PR.

minimaluminium commented on apps/admin-x-activitypub/src/components/global/ProfilePreviewHoverCard.tsx:
```diff
@@ -0,0 +1,158 @@
+        enabled: shouldFetch && Boolean(targetHandle)
+    });
+
+    const hasCompleteAccountData = accountQuery.data ? (
+        typeof accountQuery.data.followerCount === 'number' &&
+        typeof accountQuery.data.followingCount === 'number' &&
+        accountQuery.data.bio !== undefined
+    ) : false;
+
+    useEffect(() => {
+        if (!shouldFetch || !targetHandle) {
+            return;
+        }
+
+        if (!hasCompleteAccountData && !accountQuery.isFetching && !accountQuery.isLoading) {
+            accountQuery.refetch({cancelRefetch: false});
+        }
+    }, [accountQuery, accountQuery.isFetching, accountQuery.isLoading, accountQuery.refetch, hasCompleteAccountData, shouldFetch, targetHandle]);
```
The `actor` data passed to the ProfilePreviewHoverCard comes from the activity feed and contains incomplete account information (missing followerCount, followingCount, and bio).

When React Query is enabled on hover, it was returning cached incomplete data immediately and didn't make any new network request to fetch the profile data. The refetch logic ensures we make a fresh API call to the account endpoint to get complete profile data.

mike182uk replied:
> When React Query is enabled on hover, it was returning cached incomplete data immediately

huh, I'm not sure how thats possible, the query should always be returning the same data 🤔

minimaluminium replied:
I had an assumption that it was just displaying the data that came from the feed, the one that came in `author` property.

mike182uk commented on apps/admin-x-activitypub/src/components/global/ProfilePreviewHoverCard.tsx:
```diff
@@ -0,0 +1,158 @@
+        <HoverCard>
+            <HoverCardTrigger asChild>
+                <div
+                    onMouseEnter={() => {
+                        setShouldFetch(true);
+                    }}
+                    onMouseLeave={() => {
+                        setShouldFetch(false);
+                    }}
```
But lets say i'm scrolling down the feed, and my cursor hovers over every avatar as I'm scrolling down, its going to trigger the load redundantly right?

minimaluminium committed:
Refactored the preview card component

- added Account type support for the actor prop—the component will be used in various places that some of them returns ActorProperties while some passes Account type, also added a guard for safe type checking between these two types
- used existing getUsername helper instead of implementing it manually
- used onOpenChange event of the hover card component instead of mouseenter and mouseleave events to prevent accidendal network requests on hover
- fixed small ui bug that handle being not truncated in the card

minimaluminium committed:
Rebuilt yarn.lock

minimaluminium committed:
Used portal for the hover card content

The preview card content was hidden behind some elements without portal

minimaluminium committed:
Used the profile preview card component everywhere else necessary

minimaluminium committed:
Removed the feature flag

minimaluminium committed:
Fixed the placement of key attribute

minimaluminium committed:
Added asChild to hover card trigger

There were some errors when the trigger children contains same element as the default trigger of shadcn component
